File: ValidadorPolinomios/src/Polinomio/Termo.java
package Polinomio;

import java.util.ArrayList;
import java.util.HashMap;

public class Termo implements Base {
	private Sinal sinal;
	private ArrayList<Elemento> elementos;
	private Elemento ultimoElemento;
	private Expressao origem;

	public Termo() {
		elementos = new ArrayList<>();
		this.setSinal(Sinal.Positivo);
	}

	public ArrayList<Elemento> getElementos() {
		return elementos;
	}

	public void addElementos(Elemento elemento) {
		ultimoElemento = elemento;
		elemento.setOrigem(this);
		this.elementos.add(elemento);
	}

	public Elemento getUltimoElemento() {
		return ultimoElemento;
	}

	public double calcular(HashMap<Character, Integer> variaveis) {
		double result = 0;
		boolean primeiro = true;
		for (Elemento elemento : elementos) {
			if (primeiro) {
				primeiro = false;
				result = elemento.calcular(variaveis);
			} else {
				result *= elemento.calcular(variaveis);
			}
		}
		result *= sinal.getValue();
		return result;
	}

	public Expressao getOrigem() {
		return origem;
	}

	@Override
	public void setOrigem(Base origem) {
		this.origem = (Expressao) origem;
	}

	public Sinal getSinal() {
		return sinal;
	}

	public void setSinal(Sinal sinal) {
		this.sinal = sinal;
	}

	@Override
	public String toString(boolean traduzido, HashMap<Character, Integer> variaveis) {
		String result = "";
		boolean primeiro = true;
		for (Elemento elemento : elementos) {
			if (primeiro) {
				primeiro = false;
				result += elemento.toString(traduzido, variaveis);
			} else {
				result += "*" + elemento.toString(traduzido, variaveis);
			}
		}
		if (sinal == Sinal.Negativo)
			result = "(" + getSinal().toString() + result + ")";
		return result;
	}

	public Termo simplificar() {
		
		// Retira os 1 desnecessários
		// Só irá retirar se a potencia dele for um número e for positiva
		if (elementos.size() > 1) {
			Elemento elementoAjustado = elementos.get(0);
			for (int i = 0; i < elementos.size(); i++) {
				Elemento eleAux = elementos.get(i);
				if (!eleAux.temNumero() || eleAux.getNumero() != 1) {
					elementoAjustado = eleAux;
					break;
				}
			}
			
			Sinal sinalAplicar = Sinal.Positivo;
			for (int i = 0; i < elementos.size(); i++) {
				Elemento eleRemovido = elementos.get(i);
				if (eleRemovido.temNumero() && eleRemovido.getNumero() == 1) {
					if (!eleRemovido.temPotencia() || (eleRemovido.getPotencia().temNumero()
							&& eleRemovido.getPotencia().getSinal() == Sinal.Positivo)) {

						if (eleRemovido != elementoAjustado) {
							sinalAplicar = sinalAplicar.multiplicar(eleRemovido.getSinal());
							elementos.remove(i);
							i--;
						}
					}
				}
			}
			elementoAjustado.setSinal(elementoAjustado.getSinal().multiplicar(sinalAplicar));
		}
		
		// Aplica o sinal negativo do termo nos seus elementos
		if (this.getSinal() == Sinal.Negativo) {
			for (Elemento elemento : elementos) {
				elemento.setSinal(elemento.getSinal().multiplicar(this.getSinal()));
			}
			this.setSinal(this.getSinal().inverter());
		}
		
		for (Elemento elemento : elementos) {
			elemento.simplificar();
			if (elemento.temPotencia())
				elemento.getPotencia().simplificar();
		}



		if (this.elementos.size() == 1) {
			ArrayList<Termo> termosSubstituidores = new ArrayList<>();
			Elemento elementoEliminado = this.elementos.get(0);
			Expressao expressao;
			if ((expressao = elementoEliminado.getExpressao()) != null) {
				termosSubstituidores.addAll(expressao.getTermos());
				for (Termo termo : termosSubstituidores) {
					termo.setSinal(
							termo.getSinal()
							.multiplicar(this.getSinal())
							.multiplicar(elementoEliminado.getSinal()));
				}
				this.origem.substituir(this, termosSubstituidores);
			}
		} else if (this.elementos.size() > 1) {
			ArrayList<Elemento> elementosSubstituidores = new ArrayList<>();
			for (int i = 0; i < this.elementos.size(); i++) {
				Elemento elemento = elementos.get(i);
				Expressao expressao;
				if ((expressao = elemento.getExpressao()) != null) {
					if (expressao.getTermos().size() == 1) {
						Termo termoEliminado = expressao.getTermos().get(0);
						elementosSubstituidores.addAll(termoEliminado.getElementos());
						for (Elemento elementoAux : elementosSubstituidores) {
							elementoAux.setSinal(
									elementoAux.getSinal()
									.multiplicar(elemento.getSinal())
									.multiplicar(termoEliminado.getSinal()));
						}
						this.substituir(elemento, elementosSubstituidores);
						i--;
					}
				}
			}
		}

		return this;
	}

	private void substituir(Elemento elementoSubstituido, ArrayList<Elemento> elementosSubstituidores) {
		int indexElementoSubstituido = this.elementos.indexOf(elementoSubstituido);
		this.elementos.remove(indexElementoSubstituido);

		for (Elemento elementoSubstituidor : elementosSubstituidores) {
			Sinal novoSinal = 
					elementoSubstituidor.getSinal().multiplicar(elementoSubstituido.getSinal());
			elementoSubstituidor.setSinal(novoSinal);
			elementoSubstituidor.setOrigem(this);
			this.elementos.add(indexElementoSubstituido++, elementoSubstituidor);
		}

	}
}

File: ValidadorPolinomios_CS/ValidadorPolinomios/ValidadorPolinomios/Polinomio/Termo.cs
using System.Collections.Generic;

namespace Polinomio
{


	public class Termo : Base
	{
		private Sinal sinal;
		private List<Elemento> elementos;
		private Elemento ultimoElemento;
		private Expressao origem;

		public Termo()
		{
			elementos = new List<>();
			this.Sinal = Sinal.Positivo;
		}

		public virtual List<Elemento> Elementos
		{
			get
			{
				return elementos;
			}
		}

		public virtual void addElementos(Elemento elemento)
		{
			ultimoElemento = elemento;
			elemento.Origem = this;
			this.elementos.Add(elemento);
		}

		public virtual Elemento UltimoElemento
		{
			get
			{
				return ultimoElemento;
			}
		}

		public override double calcular(Dictionary<char?, int?> variaveis)
		{
			double result = 0;
			bool primeiro = true;
			foreach (Elemento elemento in elementos)
			{
				if (primeiro)
				{
					primeiro = false;
					result = elemento.calcular(variaveis);
				}
				else
				{
					result *= elemento.calcular(variaveis);
				}
			}
			result *= sinal.Value;
			return result;
		}

		public override Expressao getOrigem()
		{
			return origem;
		}

		public override void setOrigem(Base origem)
		{
			this.origem = (Expressao) origem;
		}

		public virtual Sinal Sinal
		{
			get
			{
				return sinal;
			}
			set
			{
				this.sinal = value;
			}
		}


		public override string ToString(bool traduzido, Dictionary<char?, int?> variaveis)
		{
			string result = "";
			bool primeiro = true;
			foreach (Elemento elemento in elementos)
			{
				if (primeiro)
				{
					primeiro = false;
					result += elemento.ToString(traduzido, variaveis);
				}
				else
				{
					result += "*" + elemento.ToString(traduzido, variaveis);
				}
			}
			if (sinal == Sinal.Negativo)
			{
				result = "(" + Sinal.ToString() + result + ")";
			}
			return result;
		}

		public override Termo simplificar()
		{

			// Retira os 1 desnecessários
			// Só irá retirar se a potencia dele for um número e for positiva
			if (elementos.Count > 1)
			{
				Elemento elementoAjustado = elementos[0];
				for (int i = 0; i < elementos.Count; i++)
				{
					Elemento eleAux = elementos[i];
					if (!eleAux.temNumero() || eleAux.Numero != 1)
					{
						elementoAjustado = eleAux;
						break;
					}
				}

				Sinal sinalAplicar = Sinal.Positivo;
				for (int i = 0; i < elementos.Count; i++)
				{
					Elemento eleRemovido = elementos[i];
					if (eleRemovido.temNumero() && eleRemovido.Numero == 1)
					{
						if (!eleRemovido.temPotencia() || (eleRemovido.Potencia.temNumero() && eleRemovido.Potencia.Sinal == Sinal.Positivo))
						{

							if (eleRemovido != elementoAjustado)
							{
								sinalAplicar = sinalAplicar.multiplicar(eleRemovido.Sinal);
								elementos.RemoveAt(i);
								i--;
							}
						}
					}
				}
				elementoAjustado.Sinal = elementoAjustado.Sinal.multiplicar(sinalAplicar);
			}

			// Aplica o sinal negativo do termo nos seus elementos
			if (this.Sinal == Sinal.Negativo)
			{
				foreach (Elemento elemento in elementos)
				{
					elemento.Sinal = elemento.Sinal.multiplicar(this.Sinal);
				}
				this.Sinal = this.Sinal.inverter();
			}

			foreach (Elemento elemento in elementos)
			{
				elemento.simplificar();
				if (elemento.temPotencia())
				{
					elemento.Potencia.simplificar();
				}
			}



			if (this.elementos.Count == 1)
			{
				List<Termo> termosSubstituidores = new List<Termo>();
				Elemento elementoEliminado = this.elementos[0];
				Expressao expressao;
				if ((expressao = elementoEliminado.Expressao) != null)
				{
					termosSubstituidores.AddRange(expressao.Termos);
					foreach (Termo termo in termosSubstituidores)
					{
						termo.Sinal = termo.Sinal.multiplicar(this.Sinal).multiplicar(elementoEliminado.Sinal);
					}
					this.origem.substituir(this, termosSubstituidores);
				}
			}
			else if (this.elementos.Count > 1)
			{
				List<Elemento> elementosSubstituidores = new List<Elemento>();
				for (int i = 0; i < this.elementos.Count; i++)
				{
					Elemento elemento = elementos[i];
					Expressao expressao;
					if ((expressao = elemento.Expressao) != null)
					{
						if (expressao.Termos.Count == 1)
						{
							Termo termoEliminado = expressao.Termos[0];
							elementosSubstituidores.AddRange(termoEliminado.Elementos);
							foreach (Elemento elementoAux in elementosSubstituidores)
							{
								elementoAux.Sinal = elementoAux.Sinal.multiplicar(elemento.Sinal).multiplicar(termoEliminado.Sinal);
							}
							this.substituir(elemento, elementosSubstituidores);
							i--;
						}
					}
				}
			}

			return this;
		}

		private void substituir(Elemento elementoSubstituido, List<Elemento> elementosSubstituidores)
		{
			int indexElementoSubstituido = this.elementos.IndexOf(elementoSubstituido);
			this.elementos.RemoveAt(indexElementoSubstituido);

			foreach (Elemento elementoSubstituidor in elementosSubstituidores)
			{
				Sinal novoSinal = elementoSubstituidor.Sinal.multiplicar(elementoSubstituido.Sinal);
				elementoSubstituidor.Sinal = novoSinal;
				elementoSubstituidor.Origem = this;
				this.elementos.Insert(indexElementoSubstituido++, elementoSubstituidor);
			}

		}

		public override Base ordenar()
		{
			foreach (Elemento elemento in elementos)
			{
				elemento.ordenar();
			}
			this.ordernar(elementos);
			return this;
		}

		public override int Peso
		{
			get
			{
				int total = 0;
				foreach (Elemento elemento in elementos)
				{
					total += elemento.Peso;
				}
				return total;
			}
		}
	}

}